pkg/controllers/cstorvolumeclaim/controller.go

```diff
@@ -0,0 +1,716 @@
+			c.workqueue.Add(objName)
+		}
+	*/
+}
+
+// synCVC is the function which tries to converge to a desired state for the
+// CStorVolumeClaims
+func (c *CVCController) syncCVC(cvc *apis.CStorVolumeClaim) error {
+	var err error
+	// CStor Volume Claim should be deleted. Check if deletion timestamp is set
+	// and remove finalizer.
+	if c.isClaimDeletionCandidate(cvc) {
+		klog.Infof("syncClaim: remove finalizer for CStorVolumeClaimVolume [%s]", cvc.Name)
+		err = c.removeClaimFinalizer(cvc)
+		if err != nil {
+			c.recorder.Eventf(cvc, corev1.EventTypeWarning, DeProvisioning, err.Error())
+		}
+		return nil
+	}
+
+	volName := cvc.Name
+	if volName == "" {
+		// We choose to absorb the error here as the worker would requeue the
+		// resource otherwise. Instead, the next time the resource is updated
+		// the resource will be queued again.
+		runtime.HandleError(fmt.Errorf("%+v: cvc name must be specified", cvc))
+		return nil
+	}
+
+	nodeID := cvc.Publish.NodeID
+	if nodeID == "" {
+		// We choose to absorb the error here as the worker would requeue the
+		// resource otherwise. Instead, the next time the resource is updated
+		// the resource will be queued again.
+		runtime.HandleError(fmt.Errorf("cvc must be publish/attached to Node: %+v", cvc))
+		c.recorder.Event(cvc, corev1.EventTypeWarning,
+			Provisioning,
+			fmt.Sprintf(MessageCVCPublished, cvc.Name),
+		)
+		return nil
+	}
+
+	if cvc.Status.Phase == apis.CStorVolumeClaimPhasePending {
+		klog.V(2).Infof("provisioning cstor volume %+v", cvc)
+		_, err = c.createVolumeOperation(cvc)
+		if err != nil {
+			//Record an event to indicate that any provisioning operation is failed.
+			c.recorder.Eventf(cvc, corev1.EventTypeWarning, Provisioning, err.Error())
+		}
+	}
+	// If an error occurs during Get/Create, we'll requeue the item so we can
+	// attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+	if err != nil {
+		return err
+	}
+
+	if c.cvcNeedResize(cvc) {
+		err = c.resizeCVC(cvc)
+	}
+	// If an error occurs during Get/Create, we'll requeue the item so we can
+	// attempt processing again later. This could have been caused by a
+	// temporary network failure, or any other transient reason.
+	if err != nil {
+		return err
+	}
+
+	if c.isCVCScalePending(cvc) {
+		// process scale-up/scale-down of volume replicas only if there is
+		// change in curent and desired state of replicas pool information
+		_ = c.scaleVolumeReplicas(cvc)
+	}
+	return nil
+}
+
+// UpdateCVCObj updates the cstorvolumeclaim object resource to reflect the
+// current state of the world
+func (c *CVCController) updateCVCObj(
+	cvc *apis.CStorVolumeClaim,
+	cv *apis.CStorVolume,
+) error {
+	// NEVER modify objects from the store. It's a read-only, local cache.
+	// You can use DeepCopy() to make a deep copy of original object and modify this copy
+	// Or create a copy manually for better performance
+	cvcCopy := cvc.DeepCopy()
```

Comment:
Already deep copied above ...

Reply:
will remove the extra deepcopy forget to remove during some refactor changes

Reply:
ToDO: Next PR.